File: 322-coin-change/coin-change.cpp
class Solution {
public:

    int fn(int idx, vector<int>& coins, int amount, vector<vector<int>>& dp)
    {
        if(amount == 0) return 0;
        if(amount < 0 || idx == coins.size()) return INT_MAX; 
        if(dp[idx][amount] != -1 )return dp[idx][amount];
         int take = fn(idx, coins, amount - coins[idx],dp);
        if(take != INT_MAX) take += 1;
        int notTake = fn(idx + 1, coins, amount,dp);

        return dp[idx][amount] = min(take, notTake);


    }



    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(amount + 1, -1));
        int ans= fn(0, coins, amount,dp);
        return ans == INT_MAX ? -1 :ans;
    }
};

File: 322-coin-change/coin-change.java
class Solution {

    private int fn(int idx, int[] coins, int amount,int[][] dp){

        if(idx == 0)
        {
            if(amount % coins[idx] == 0) return amount / coins[idx];
            else return (int)1e9;
        }

        if(dp[idx][amount] != -1) return dp[idx][amount];

        int notPick = fn(idx - 1, coins, amount,dp);
        int pick = (int)1e9;
        if(amount >= coins[idx])
        {
            pick = 1 + fn(idx , coins, amount - coins[idx],dp);
        }
        dp[idx][amount] = Math.min(pick , notPick);
        return dp[idx][amount];
    }
    public int coinChange(int[] coins, int amount) {
        int[][]dp =new int[coins.length][amount+1];
        for(int[] arr :dp)
        Arrays.fill(arr, -1);
        int ans = fn(coins.length - 1, coins, amount,dp);
        return (ans == (int)1e9) ? -1 : ans;
    }
}